osu.Server.Spectator/Hubs/ScoreProcessedSubscriber.cs

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using osu.Framework.Logging;
+using osu.Server.Spectator.Database;
+using StackExchange.Redis;
+using StatsdClient;
+using Timer = System.Timers.Timer;
+
+namespace osu.Server.Spectator.Hubs;
+
+public sealed class ScoreProcessedSubscriber : IScoreProcessedSubscriber, IDisposable
+{
+    /// <summary>
+    /// The maximum amount of time to wait for a <see cref="ScoreProcessed"/> message for a given score in milliseconds.
+    /// </summary>
+    private const int timeout_interval_ms = 30_000;
+
+    private const string statsd_prefix = "subscribers.score-processed";
+
+    private readonly IDatabaseFactory databaseFactory;
+    private readonly ISubscriber? subscriber;
+
+    private readonly ConcurrentDictionary<long, ScoreProcessedSubscription> subscriptions = new ConcurrentDictionary<long, ScoreProcessedSubscription>();
+    private readonly Timer timer;
+    private readonly Logger logger;
+
+    public ScoreProcessedSubscriber(
+        IDatabaseFactory databaseFactory,
+        IConnectionMultiplexer redis)
+    {
+        this.databaseFactory = databaseFactory;
+
+        timer = new Timer(1000);
+        timer.AutoReset = true;
+        timer.Elapsed += (_, _) => Task.Run(purgeTimedOutSubscriptions);
+        timer.Start();
+
+        subscriber = redis.GetSubscriber();
+        subscriber.Subscribe("osu-channel:score:processed", (_, message) => onMessageReceived(message));
+
+        logger = Logger.GetLogger(nameof(ScoreProcessedSubscriber));
+    }
+
+    private void onMessageReceived(string message)
+    {
+        try
+        {
+            var scoreProcessed = JsonConvert.DeserializeObject<ScoreProcessed>(message);
+
+            if (scoreProcessed == null)
+                return;
+
+            if (subscriptions.TryRemove(scoreProcessed.ScoreId, out var subscription))
+            {
+                using (subscription)
+                    subscription.InvokeAsync().Wait();
```

Comment:
The `.Wait()` is unfortunate, but I'm not sure how to avoid it.

Reply:
Probably fine.